brave/brave-ios
# Fix #6921: Persist Asset Balance in Portfolio

## Summary of Changes
We are now caching asset's balance. They can be easily fetch from CD via token info or account info or both. 
Cached balance will be refreshed under certain scenarios. they will be listed in the test plan.


This pull request fixes brave/brave-browser#35986

## Submitter Checklist:


## Test Plan:

- Upgrade (current target 1.63.x)
1. install a older version of Brave (1.62.x and earlier)
2. set up a wallet with some balance 
3. Dismiss wallet then open 
4. Notice every time wallet open triggers balance fetching
5. upgrade Brave to the 1.63.x
6. Open wallet you will notice there will be a loading time to fetch balance since we have never cached any
7. Dismiss wallet and open again
8. Notice the balance still being displayed with no loading. 

- New Wallet Set up
1. Fresh install wallet and create a new wallet from onboarding 
9. All default user assets will appear in Portfolio with 0 balance. 
10. Dismiss the wallet then open it (without locking)
11. There will be no loading in Portfolio. 

- Restore A Wallet 
Note: Currently, we have known issues https://github.com/brave/brave-browser/issues/35696 and https://github.com/brave/brave-browser/issues/35697 these two cause auto-discovery assets are not discovered right after restoring. We will use a workaround to verify auto-discovery assets can be discovered and their balances are being fetched correctly. But this issue helps us to verify another scenario I will explain later in other test steps.
1. Delete Brave and fresh install Brave 
2. Restore a wallet from onboarding (Wallet A)
3. You will see default user assets are being displayed and their balance are being fetched or in within couple seconds based on internet speed.
4. Dismiss the wallet then open it (without locking)
12. There will be no loading in Portfolio.
13. Notice there is no auto-discovered assets got discovered. This is the issue mentioned earlier. 
14. Let's take advantage of this issue by checking if wallet will add and fetch balance of a token that we already know this wallet has
15. Let's say we already know Wallet A has some balance of `WETH` on `Polygon Mainnet`
16. Tap Edit Visible Asset to bring up the token list, and search for `WETH`
17. Tap on `WETH` to mark it as visible asset
18. Back to Portfolio, check `WETH` now appears and its balance is fetched. 
19. A workaround to let core give the rest of auto-discovery assets is go to `NFTs` tab, enable NFT auto-discovery in the prompt
20. A spinner will show up indicating wallet starts discovering assets. 
21. The spinner under `Assets` also shows up if you navigate to `Assets`
22. After a while, once the spinner disappears, some new assets will show up in `Assets` that this Wallet A owns with some balance. Please check if their balances were fetched correctly. 
23. repeat step 4-5
24. Lock the wallet
25. restore a new wallet (Wallet A)
26. Check user assets and balance are fetched correctly, 
27. Repeat step 4-5
- Add a primary or secondary account to wallet 
1. Group visible assets by `Account` in Portfolio 
2. add a primary account 
3. check the native token has been added to Portfolio and its balance is fetched "0"
4. add a secondary account with some balance 
6. check all visible assets have the same coin type are displayed under this new account and their balances are fetched correctly 
- Add/Remove a new network 
1. Group visible assets by `Group`
2. add/remove a custom network from wallet settings or a dapp 
3. Check if the new network group is added in Portfolio and its native asset is displayed in the group with its balance fetched correctly. (Or removed if the network is removed)
- Add/Remove custom token
1. add/remove a custom token from wallet settings or a dapp 
2. Check if the new token is displayed in Portfolio with its balance fetched correctly. (Or removed if the token is removed)
- Check balance after a transaction is confirmed/error/dropped 
1. make a send transaction of token A from Portfolio 
2. after the transaction status becomes `Confirmed`
3. check if token A balance is updated
4. Not sure how to make a transaction error out or dropped. but Once a transaction status becomes error or dropped. Some gas fee was consumed so the native token balance is relatively decreased. 

In summary, balances in Portfolio (Assets) will be refreshed when
1. A wallet is created 
2. A wallet is restored
3. When wallet is unlocked 
4. A account is created or added 
6. A network is added
7. An asset is added 
8. An asset's visibility status is changed to be visible
9. A transaction is confirmed or error out or dropped

Note: Transactions that are submitted by Dapps via signing request through wallet, since there is no access from Wallet to know the transaction status that is not created by Brave Wallet, it is not possible to determine when to refresh the balance based this transaction's status until we can observe all on-chain activity for an address. In this case, to refresh balance, user has to lock and unlock the wallet. 

## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `bug` / `enhancement`
